plugin/data-manipulation/data.manipulation.view.js

```diff
@@ -0,0 +1,234 @@
+					edited.set(rowId, entries);
+				}
+
+				let entryIndex = entries.findIndex(entry => entry.column === e.column.key);
+				let entry = entries[entryIndex];
+				if (!entry) {
+					entry = {
+						column: e.column.key,
+						oldValue: e.oldValue,
+						oldLabel: e.oldLabel
+					};
+
+					entryIndex = entries.length;
+					entries.push(entry);
+				}
+
+				entry.newValue = e.newValue;
+				entry.newLabel = e.newLabel;
+
+				// TODO: understand if we need to track label changes?
+				if (!this.hasChanges(entry.newValue, entry.oldValue)) {
+					entries.splice(entryIndex, 1);
+					if (!entries.length) {
+						edited.delete(rowId);
+					}
+				}
+			}
+		});
+
+		this.actions = [
+			new Action(
+				new Command({
+					source: 'data.manipulation',
+					execute: () => {
+						const newRow = this.rowFactory(this.model.data().rows[0]);
+						if (isUndefined(newRow)) {
+							throw new AppError('data.manipulation', 'Setup rowFactory property to add new rows');
+						}
+
+						const rowId = this.rowId(newRow);
+						const data = this.model.data;
+
+						this.changes.added.add(rowId);
+						data({
+							rows: [newRow].concat(data().rows)
+						});
+					},
+					shortcut: 'F7'
+				}),
+				'Add New Row',
+				'add'
+			)];
+
+		this.rowActions = [
+			new Action(
+				new Command({
+					source: 'data.manipulation',
+					canExecute: e => {
+						const rowId = this.rowId(e.row);
+						return !this.changes.deleted.has(rowId);
+					},
+					execute: e => {
+						const rowId = this.rowId(e.row);
+						const changes = this.changes;
+						if (changes.added.has(rowId)) {
+							changes.added.delete(rowId);
+							const data = this.model.data;
+							const rows = data().rows.filter(row => this.rowId(row) !== rowId);
+							data({rows});
+						}
+						else {
+							changes.deleted.add(rowId);
+						}
+					}
+				}),
+				'Delete Row',
+				'delete'
+			),
+			new Action(
+				new Command({
+					source: 'data.manipulation',
+					execute: e => {
+						const rowId = this.rowId(e.row);
+						if (this.changes.deleted.has(rowId)) {
+							this.changes.deleted.delete(rowId);
+						}
+
+						if (this.changes.edited.has(rowId)) {
+							try {
+								const edits = this.changes.edited.get(rowId);
+								const columnMap = columnService.map(this.model.data().columns);								
+								for(const edit of edits) {
+									const column = columnMap[edit.column];
+									if (!column){
+										throw new AppError('data.manipulation', `Column ${edit.column} is not found`);
+									}
+	
+									setValue(e.row, column, edit.oldValue);
+									setLabel(e.row, column, edit.oldLabel);
+								}
+							}
+							finally {
+								this.changes.edited.delete(rowId);	
+							}
+						}
+					},
+					canExecute: e => {
+						const rowId = this.rowId(e.row);
+						return this.changes.deleted.has(rowId) || this.changes.edited.has(rowId);
+					}
+				}),
+				'Revert Row',
+				'restore'
+			),
+			// new Action(
```

Comment:
Is this something we need?

Reply:
Yes we need to have a service that will allow to open edit form from the user code